File: NucleotideTranslator/TranslateNucleotides.vb
Public Class clsTranslateNucleotides

    Protected m_TranslationMatrix As ArrayList
    Protected m_GetSQLData As TableManipulationBase.clsDBTask

    Protected m_TranTableListName As String = "T_DNA_Translation_Tables"
    Protected m_TransTableMembersName As String = "T_DNA_Translation_Table_Members"

    Public Sub New(DMSConnectionString As String)
        Me.m_GetSQLData = New TableManipulationBase.clsDBTask(DMSConnectionString)
    End Sub

    Public Function LoadTransMatrix(TranslationTableID As Integer) As ArrayList

        Dim BaseArray() As Char = "ATGC".ToCharArray
        Dim base_1 As Char
        Dim base_2 As Char
        Dim base_3 As Char

        Dim selectSQL As String =
            "SELECT * FROM " & Me.m_TransTableMembersName &
            " WHERE DNA_Translation_Table_ID = " & TranslationTableID

        Dim members As DataTable = Me.m_GetSQLData.GetTable(selectSQL)

        Dim dr As DataRow
        Dim tertSelect As String
        Dim TertiaryRows() As DataRow

        Dim PrimaryList As New ArrayList
        Dim SecondaryList As New ArrayList
        Dim TertiaryList As New ArrayList

        For Each base_1 In BaseArray
            For Each base_2 In BaseArray

                For Each base_3 In BaseArray
                    tertSelect = "Base_1 = '" & base_1.ToString &
                        "' AND Base_2 = '" & base_2.ToString &
                        "' AND Base_3 = '" & base_3.ToString & "'"
                    TertiaryRows = members.Select(tertSelect)

                    dr = TertiaryRows(0)

                    TertiaryList.Add(New clsTranslationEntry(base_3.ToString, CStr(dr.Item("Coded_AA"))))
                Next
                SecondaryList.Add(New clsTranslationEntry(base_2.ToString, TertiaryList))
                TertiaryList = New ArrayList
            Next

            PrimaryList.Add(New clsTranslationEntry(base_1.ToString, SecondaryList))
            SecondaryList = New ArrayList
        Next

        Return PrimaryList
    End Function

    Protected Function LoadNucPositions(filePath As String) As Integer

    End Function

End Class

File: NucleotideTranslator/TranslateNucleotides.cs
using System.Collections;
using System.Data;

namespace NucleotideTranslator
{
    public class TranslateNucleotides
    {
        private ArrayList m_TranslationMatrix;
        private TableManipulationBase.DBTask m_GetSQLData;

        private string m_TranTableListName = "T_DNA_Translation_Tables";
        private string m_TransTableMembersName = "T_DNA_Translation_Table_Members";

        public TranslateNucleotides(string DMSConnectionString)
        {
            m_GetSQLData = new TableManipulationBase.DBTask(DMSConnectionString);
        }

        public ArrayList LoadTransMatrix(int TranslationTableID)
        {
            var BaseArray = "ATGC".ToCharArray();

            string selectSQL =
                "SELECT * FROM " + m_TransTableMembersName +
                " WHERE DNA_Translation_Table_ID = " + TranslationTableID;

            var members = m_GetSQLData.GetTable(selectSQL);

            DataRow dr;
            string tertSelect;
            DataRow[] TertiaryRows;

            var PrimaryList = new ArrayList();
            var SecondaryList = new ArrayList();
            var TertiaryList = new ArrayList();

            foreach (var base_1 in BaseArray)
            {
                foreach (var base_2 in BaseArray)
                {
                    foreach (var base_3 in BaseArray)
                    {
                        tertSelect = "Base_1 = '" + base_1.ToString() +
                            "' AND Base_2 = '" + base_2.ToString() +
                            "' AND Base_3 = '" + base_3.ToString() + "'";
                        TertiaryRows = members.Select(tertSelect);

                        dr = TertiaryRows[0];

                        TertiaryList.Add(new TranslationEntry(base_3.ToString(), dr["Coded_AA"].ToString()));
                    }

                    SecondaryList.Add(new TranslationEntry(base_2.ToString(), TertiaryList));
                    TertiaryList = new ArrayList();
                }

                PrimaryList.Add(new TranslationEntry(base_1.ToString(), SecondaryList));
                SecondaryList = new ArrayList();
            }

            return PrimaryList;
        }

        protected int LoadNucPositions(string filePath)
        {
            return default;
        }
    }
}